File: app/src/main/java/com/bsoft/inventorymanager/viewmodels/CreatePurchaseViewModel.java
package com.bsoft.inventorymanager.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.bsoft.inventorymanager.models.Product;
import com.bsoft.inventorymanager.models.ProductSelection;
import com.bsoft.inventorymanager.models.Purchase;
import com.bsoft.inventorymanager.models.PurchaseItem;
import com.bsoft.inventorymanager.model.Supplier;
import com.bsoft.inventorymanager.repositories.PurchaseRepository;
import com.bsoft.inventorymanager.utils.FinancialCalculator;
import com.google.firebase.Timestamp;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class CreatePurchaseViewModel extends ViewModel {

    private final PurchaseRepository purchaseRepository;

    private final MutableLiveData<List<ProductSelection>> productSelections = new MutableLiveData<>(new ArrayList<>());
    private final MutableLiveData<Double> subtotal = new MutableLiveData<>(0.0);
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    private final MutableLiveData<String> error = new MutableLiveData<>();
    private final MutableLiveData<String> purchaseSuccess = new MutableLiveData<>();
    private final MutableLiveData<List<Supplier>> suppliers = new MutableLiveData<>();
    private final MutableLiveData<Supplier> selectedSupplier = new MutableLiveData<>();

    @Inject
    public CreatePurchaseViewModel(PurchaseRepository purchaseRepository) {
        this.purchaseRepository = purchaseRepository;
        loadSuppliers();
    }

    public LiveData<List<ProductSelection>> getProductSelections() {
        return productSelections;
    }

    public LiveData<Double> getSubtotal() {
        return subtotal;
    }

    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    public LiveData<String> getError() {
        return error;
    }

    public LiveData<String> getPurchaseSuccess() {
        return purchaseSuccess;
    }

    public LiveData<List<Supplier>> getSuppliers() {
        return suppliers;
    }

    public LiveData<Supplier> getSelectedSupplier() {
        return selectedSupplier;
    }

    public void loadSuppliers() {
        isLoading.setValue(true);
        purchaseRepository.getSuppliers(new PurchaseRepository.SuppliersCallback() {
            @Override
            public void onSuccess(List<Supplier> supplierList) {
                isLoading.setValue(false);
                suppliers.setValue(supplierList);
            }

            @Override
            public void onFailure(Exception e) {
                isLoading.setValue(false);
                error.setValue("Failed to load suppliers: " + e.getMessage());
            }
        });
    }

    public void setSelectedSupplier(Supplier supplier) {
        selectedSupplier.setValue(supplier);
    }

    public void addProduct(Product product) {
        if (product == null)
            return;

        List<ProductSelection> currentItems = productSelections.getValue();
        if (currentItems == null)
            currentItems = new ArrayList<>();

        boolean found = false;
        for (int i = 0; i < currentItems.size(); i++) {
            if (currentItems.get(i).getProduct().getDocumentId().equals(product.getDocumentId())) {
                ProductSelection matchingItem = currentItems.get(i);
                matchingItem.setQuantityInSale(matchingItem.getQuantityInSale() + 1);
                found = true;
                break;
            }
        }

        if (!found) {
            currentItems.add(new ProductSelection(product, 1));
        }

        productSelections.setValue(currentItems);
        recalculateSubtotal();
    }

    public void loadProductByBarcode(String barcode) {
        isLoading.setValue(true);
        purchaseRepository.getProductByBarcode(barcode, new PurchaseRepository.ProductCallback() {
            @Override
            public void onSuccess(Product product) {
                isLoading.setValue(false);
                addProduct(product);
            }

            @Override
            public void onFailure(Exception e) {
                isLoading.setValue(false);
                error.setValue("Product not found or error: " + e.getMessage());
            }
        });
    }

    public void removeProduct(int position) {
        List<ProductSelection> currentItems = productSelections.getValue();
        if (currentItems != null && position >= 0 && position < currentItems.size()) {
            currentItems.remove(position);
            productSelections.setValue(currentItems);
            recalculateSubtotal();
        }
    }

    public void updateItemQuantity(int position, int quantity) {
        List<ProductSelection> currentItems = productSelections.getValue();
        if (currentItems != null && position >= 0 && position < currentItems.size()) {
            ProductSelection item = currentItems.get(position);
            item.setQuantityInSale(quantity);
            // CRITICAL FIX: Do NOT call productSelections.setValue(currentItems) here.
            // This prevents the Adapter from refreshing/losing focus while typing.
            recalculateSubtotal();
        }
    }

    public void updateItemPrice(int position, double price) {
        List<ProductSelection> currentItems = productSelections.getValue();
        if (currentItems != null && position >= 0 && position < currentItems.size()) {
            ProductSelection item = currentItems.get(position);
            item.getProduct().setPurchasePrice(price);
            // CRITICAL FIX: Do NOT call productSelections.setValue(currentItems) here.
            recalculateSubtotal();
        }
    }

    private void recalculateSubtotal() {
        double sub = 0.0;
        List<ProductSelection> items = productSelections.getValue();
        if (items != null) {
            for (ProductSelection item : items) {
                sub += FinancialCalculator.calculateLineItemTotal(item.getProduct().getPurchasePrice(),
                        item.getQuantityInSale());
            }
        }
        subtotal.setValue(sub);
    }

    public void savePurchase(Date purchaseDate, double taxAmount, double discountAmount, String paymentMethod,
            double amountPaid, String notes) {
        if (isLoading.getValue() == Boolean.TRUE)
            return;

        List<ProductSelection> selections = productSelections.getValue();
        if (selections == null || selections.isEmpty()) {
            error.setValue("No products selected");
            return;
        }

        if (selectedSupplier.getValue() == null) {
            error.setValue("No supplier selected");
            return;
        }

        isLoading.setValue(true);

        Purchase purchase = new Purchase();
        purchase.setSupplierId(selectedSupplier.getValue().getDocumentId());
        purchase.setSupplierName(selectedSupplier.getValue().getName());
        purchase.setSupplierContactNumber(selectedSupplier.getValue().getContactNumber());

        Date dateParam = purchaseDate != null ? purchaseDate : new Date();
        purchase.setPurchaseDate(new Timestamp(dateParam));

        List<PurchaseItem> items = new ArrayList<>();
        List<String> productIds = new ArrayList<>();
        for (ProductSelection sel : selections) {
            PurchaseItem item = new PurchaseItem();
            item.setProductId(sel.getProduct().getDocumentId());
            item.setProductName(sel.getProduct().getName());
            item.setPricePerItem(sel.getProduct().getPurchasePrice());
            item.setQuantity(sel.getQuantityInSale());
            items.add(item);
            productIds.add(sel.getProduct().getDocumentId());
        }
        purchase.setItems(items);
        purchase.setProductIds(productIds);

        double sub = subtotal.getValue() != null ? subtotal.getValue() : 0.0;
        // Purchase model doesn't strictly have subtotal field exposed in setters
        // usually seen in CreatePurchaseActivity
        // but let's check Purchase model. It has TotalAmount.

        purchase.setTaxAmount(taxAmount);
        purchase.setDiscountAmount(discountAmount);
        purchase.setTotalAmount(FinancialCalculator.calculateTotalAmount(sub, taxAmount, discountAmount));
        purchase.setPaymentMethod(paymentMethod);
        purchase.setAmountPaid(amountPaid);
        purchase.setNotes(notes);
        purchase.setAmountDue(purchase.getTotalAmount() - amountPaid);
        purchase.setUserId("CURRENT_USER_ID"); // TODO: Real user
        purchase.setStatus("COMPLETED");

        purchaseRepository.savePurchase(purchase, items, new PurchaseRepository.PurchaseCallback() {
            @Override
            public void onSuccess(String purchaseId) {
                isLoading.setValue(false);
                purchaseSuccess.setValue(purchaseId);
            }

            @Override
            public void onFailure(Exception e) {
                isLoading.setValue(false);
                error.setValue("Failed to save purchase: " + e.getMessage());
            }
        });
    }

    public void loadPurchaseForEditing(String purchaseId) {
        isLoading.setValue(true);
        purchaseRepository.getPurchase(purchaseId, new PurchaseRepository.GetPurchaseCallback() {
            @Override
            public void onSuccess(Purchase purchase) {
                // Populate selections
                List<ProductSelection> recoveredSelections = new ArrayList<>();
                List<PurchaseItem> items = purchase.getItems();

                // We need to fetch full product details for each item to have complete objects
                // For simplicity/speed in MVP refactor, we can fetch them individually or use
                // what we have.
                // The original Activity fetched them.
                // Let's iterate and fetch.

                // For now, load simple placeholders and trigger async fetch if needed.
                // Or better, let's use a recursive fetch or standard loop with counter like
                // usage in Activity.
                // But doing async inside VM is better handled with Coroutines. Without them, we
                // callback hell.

                // Simplification: Just load what we have from PurchaseItem into Product
                // placeholder
                // and maybe fetch details if deemed critical (e.g. current stock).
                // Original activity fetched details to get current stock and other metadata.

                reconstructSelections(items);

                // Load supplier
                loadSupplierById(purchase.getSupplierId());
                isLoading.setValue(false);
            }

            @Override
            public void onFailure(Exception e) {
                isLoading.setValue(false);
                error.setValue("Failed to load purchase: " + e.getMessage());
            }
        });
    }

    private void loadSupplierById(String supplierId) {
        // We already have suppliers list loaded?
        List<Supplier> currentList = suppliers.getValue();
        if (currentList != null) {
            for (Supplier s : currentList) {
                if (s.getDocumentId().equals(supplierId)) {
                    selectedSupplier.setValue(s);
                    return;
                }
            }
        }
        // If not in list (e.g. pagination or inactive), we might need direct fetch.
        // For now rely on list.
    }

    private void reconstructSelections(List<PurchaseItem> items) {
        // Implementation note: Ideally we fetch latest product info.
        // For this refactor, we will map PurchaseItem to ProductSelection directly.
        List<ProductSelection> list = new ArrayList<>();
        for (PurchaseItem item : items) {
            Product p = new Product();
            p.setDocumentId(item.getProductId());
            p.setName(item.getProductName());
            p.setPurchasePrice(item.getPricePerItem());
            // Quantity? We don't know current stock without fetching.
            // We set it to 0 or leave it.

            ProductSelection sel = new ProductSelection(p, item.getQuantity());
            list.add(sel);
        }
        productSelections.setValue(list);
        recalculateSubtotal();
    }
}

File: app/src/main/java/com/bsoft/inventorymanager/viewmodels/CreatePurchaseViewModel.kt
package com.bsoft.inventorymanager.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.bsoft.inventorymanager.models.Product
import com.bsoft.inventorymanager.models.ProductSelection
import com.bsoft.inventorymanager.models.Purchase
import com.bsoft.inventorymanager.models.PurchaseItem
import com.bsoft.inventorymanager.model.Supplier
import com.bsoft.inventorymanager.repositories.PurchaseRepository
import com.bsoft.inventorymanager.utils.FinancialCalculator
import com.google.firebase.Timestamp
import dagger.hilt.android.lifecycle.HiltViewModel
import java.util.Date
import javax.inject.Inject

@HiltViewModel
class CreatePurchaseViewModel @Inject constructor(
    private val purchaseRepository: PurchaseRepository
) : ViewModel() {

    private val _productSelections = MutableLiveData<List<ProductSelection>>(ArrayList())
    val productSelections: LiveData<List<ProductSelection>> = _productSelections

    private val _subtotal = MutableLiveData(0.0)
    val subtotal: LiveData<Double> = _subtotal

    private val _isLoading = MutableLiveData(false)
    
    @get:JvmName("getIsLoading")
    val isLoading: LiveData<Boolean> = _isLoading

    private val _error = MutableLiveData<String?>()
    val error: LiveData<String?> = _error

    private val _purchaseSuccess = MutableLiveData<String?>()
    val purchaseSuccess: LiveData<String?> = _purchaseSuccess

    private val _suppliers = MutableLiveData<List<Supplier>>()
    val suppliers: LiveData<List<Supplier>> = _suppliers

    private val _selectedSupplier = MutableLiveData<Supplier?>()
    val selectedSupplier: LiveData<Supplier?> = _selectedSupplier

    init {
        loadSuppliers()
    }

    fun loadSuppliers() {
        _isLoading.value = true
        purchaseRepository.getSuppliers(object : PurchaseRepository.SuppliersCallback {
            override fun onSuccess(supplierList: List<Supplier>) {
                _isLoading.value = false
                _suppliers.value = supplierList
            }

            override fun onFailure(e: Exception) {
                _isLoading.value = false
                _error.value = "Failed to load suppliers: ${e.message}"
            }
        })
    }

    fun setSelectedSupplier(supplier: Supplier?) {
        _selectedSupplier.value = supplier
    }

    fun addProduct(product: Product?) {
        if (product == null) return

        val currentItems = _productSelections.value?.toMutableList() ?: mutableListOf()

        var found = false
        for (item in currentItems) {
            if (item.product.documentId == product.documentId) {
                item.quantityInSale = item.quantityInSale + 1
                found = true
                break
            }
        }

        if (!found) {
            currentItems.add(ProductSelection(product, 1))
        }

        _productSelections.value = currentItems
        recalculateSubtotal()
    }

    fun loadProductByBarcode(barcode: String) {
        _isLoading.value = true
        purchaseRepository.getProductByBarcode(barcode, object : PurchaseRepository.ProductCallback {
            override fun onSuccess(product: Product) {
                _isLoading.value = false
                addProduct(product)
            }

            override fun onFailure(e: Exception) {
                _isLoading.value = false
                _error.value = "Product not found or error: ${e.message}"
            }
        })
    }

    fun removeProduct(position: Int) {
        val currentItems = _productSelections.value?.toMutableList()
        if (currentItems != null && position >= 0 && position < currentItems.size) {
            currentItems.removeAt(position)
            _productSelections.value = currentItems
            recalculateSubtotal()
        }
    }

    fun updateItemQuantity(position: Int, quantity: Int) {
        val currentItems = _productSelections.value
        if (currentItems != null && position >= 0 && position < currentItems.size) {
            val item = currentItems[position]
            item.quantityInSale = quantity
            // CRITICAL FIX: Do NOT call productSelections.setValue(currentItems) here.
            // This prevents the Adapter from refreshing/losing focus while typing.
            recalculateSubtotal()
        }
    }

    fun updateItemPrice(position: Int, price: Double) {
        val currentItems = _productSelections.value
        if (currentItems != null && position >= 0 && position < currentItems.size) {
            val item = currentItems[position]
            item.product.purchasePrice = price
            // CRITICAL FIX: Do NOT call productSelections.setValue(currentItems) here.
            recalculateSubtotal()
        }
    }

    private fun recalculateSubtotal() {
        var sub = 0.0
        val items = _productSelections.value
        if (items != null) {
            for (item in items) {
                sub += FinancialCalculator.calculateLineItemTotal(
                    item.product.purchasePrice,
                    item.quantityInSale
                )
            }
        }
        _subtotal.value = sub
    }

    fun savePurchase(
        purchaseDate: Date?,
        taxAmount: Double,
        discountAmount: Double,
        paymentMethod: String,
        amountPaid: Double,
        notes: String
    ) {
        if (_isLoading.value == true) return

        val selections = _productSelections.value
        if (selections == null || selections.isEmpty()) {
            _error.value = "No products selected"
            return
        }

        if (_selectedSupplier.value == null) {
            _error.value = "No supplier selected"
            return
        }

        _isLoading.value = true

        val purchase = Purchase()
        purchase.supplierId = _selectedSupplier.value!!.documentId
        purchase.supplierName = _selectedSupplier.value!!.name
        purchase.supplierContactNumber = _selectedSupplier.value!!.contactNumber

        val dateParam = purchaseDate ?: Date()
        purchase.purchaseDate = Timestamp(dateParam)

        val items = mutableListOf<PurchaseItem>()
        val productIds = mutableListOf<String>()
        for (sel in selections) {
            val item = PurchaseItem()
            item.productId = sel.product.documentId
            item.productName = sel.product.name
            item.pricePerItem = sel.product.purchasePrice
            item.quantity = sel.quantityInSale
            items.add(item)
            productIds.add(sel.product.documentId)
        }
        purchase.items = items
        purchase.productIds = productIds

        val sub = _subtotal.value ?: 0.0

        purchase.taxAmount = taxAmount
        purchase.discountAmount = discountAmount
        purchase.totalAmount = FinancialCalculator.calculateTotalAmount(sub, taxAmount, discountAmount)
        purchase.paymentMethod = paymentMethod
        purchase.amountPaid = amountPaid
        purchase.notes = notes
        purchase.amountDue = purchase.totalAmount - amountPaid
        purchase.userId = "CURRENT_USER_ID" // TODO: Real user
        purchase.status = "COMPLETED"

        purchaseRepository.savePurchase(purchase, items, object : PurchaseRepository.PurchaseCallback {
            override fun onSuccess(purchaseId: String) {
                _isLoading.postValue(false)
                _purchaseSuccess.postValue(purchaseId)
            }

            override fun onFailure(e: Exception) {
                _isLoading.postValue(false)
                _error.postValue("Failed to save purchase: ${e.message}")
            }
        })
    }

    fun loadPurchaseForEditing(purchaseId: String) {
        _isLoading.value = true
        purchaseRepository.getPurchase(purchaseId, object : PurchaseRepository.GetPurchaseCallback {
            override fun onSuccess(purchase: Purchase) {
                val items = purchase.items
                reconstructSelections(items)
                loadSupplierById(purchase.supplierId)
                _isLoading.value = false
            }

            override fun onFailure(e: Exception) {
                _isLoading.value = false
                _error.value = "Failed to load purchase: ${e.message}"
            }
        })
    }

    private fun loadSupplierById(supplierId: String) {
        val currentList = _suppliers.value
        if (currentList != null) {
            for (s in currentList) {
                if (s.documentId == supplierId) {
                    _selectedSupplier.value = s
                    return
                }
            }
        }
        // If not in list (e.g. pagination or inactive), we might need direct fetch.
        // For now rely on list.
    }

    private fun reconstructSelections(items: List<PurchaseItem>) {
        val list = mutableListOf<ProductSelection>()
        for (item in items) {
            val p = Product()
            p.documentId = item.productId
            p.name = item.productName
            p.purchasePrice = item.pricePerItem

            val sel = ProductSelection(p, item.quantity)
            list.add(sel)
        }
        _productSelections.value = list
        recalculateSubtotal()
    }
}
